code/lib/manager-api/src/modules/stories.ts

```diff
@@ -600,13 +601,25 @@ export const init: ModuleFn<SubAPI, SubState, true> = ({
         const { sourceType } = getEventMetadata(this, fullAPI);
 
         if (sourceType === 'local') {
-          if (fullAPI.isSettingsScreenActive()) return;
-
-          // Special case -- if we are already at the story being specified (i.e. the user started at a given story),
-          // we don't need to change URL. See https://github.com/storybookjs/storybook/issues/11677
           const state = store.getState();
-          if (state.storyId !== storyId || state.viewMode !== viewMode) {
-            navigate(`/${viewMode}/${storyId}`);
+          /**
+           * When storybook starts, we want to navigate to the first story.
+           * But there are a few exceptions:
+           * - If the current storyId and viewMode are already set/correct.
+           * - If the user has navigated away already.
+           * - If the user started storybook with a specific page-URL like "/settings/about"
+           */
+          if (state.path === '/' || state.viewMode === 'story' || state.viewMode === 'docs') {
+            if (
+              state.viewMode &&
+              state.storyId &&
+              state.viewMode !== viewMode &&
+              state.storyId !== storyId
+            ) {
+              fullAPI.emit(SET_CURRENT_STORY, { storyId: state.storyId, viewMode: state.viewMode });
+            } else if (state.storyId !== storyId || state.viewMode !== viewMode) {
+              navigate(`/${viewMode}/${storyId}`);
+            }
```

Comment:
I think this code could be a lot clearer if some of these checks were pulled out into variables, like:

```
const stateHasSelection = state.viewMode && state.storyId;
const stateSelectionDifferent = state.storyId !== storyId || state.viewMode !== viewMode;
```